runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/functions/GroupAlsoByWindowViaOutputBufferFn.java

```diff
@@ -65,9 +65,15 @@ public GroupAlsoByWindowViaOutputBufferFn(
 
   @Override
   public Iterator<WindowedValue<KV<K, Iterable<InputT>>>> call(
-      KV<K, Iterable<WindowedValue<InputT>>> kv) throws Exception {
-    K key = kv.getKey();
-    Iterable<WindowedValue<InputT>> values = kv.getValue();
+      K key, Iterator<WindowedValue<KV<K, InputT>>> iterator) throws Exception {
+
+    // we have to meterialize the Iterator because ReduceFnRunner.processElements expects
+    // ArrayList<WindowedValue<InputT>> and not Iterator<WindowedValue<KV<K, InputT>>>
+    ArrayList<WindowedValue<InputT>> values = new ArrayList<>();
+    while (iterator.hasNext()) {
+      WindowedValue<KV<K, InputT>> wv = iterator.next();
+      values.add(wv.withValue(wv.getValue().getValue()));
```

Comment:
I think we still can have potential OOM here, right?

Reply:
Yes this comment in the doc does seem to confirm this '...users must take care to avoid materializing the whole iterator for a group (for example, by calling toList) unless they are sure that this is possible given the memory constraints of their cluster.` this looks like doing exactly the same than I would expect the `toList` call to do.

Reply:
I know but as stated in my comment in the code, it is unavoidable due to the reduceFnRunner needing a list as input.
Anyway I measured during the load test (see results above) in a tiny JVM and I got no OOM, I only got spill to disc of GB of data

Reply:
Well, I think it depends on values distribution over keys. OOM can happened here on hot keys with many or/and large values. You can try to avoid it using `Iterators.transform()` and creating `Iterable` from this (should be checked if it will work).

Reply:
Yes sure, I agree about hot keys. But if spark spills to disk it is because it is out of memory. And this PR does not claim to fix the OOM issue that also existed in the previous version of GBK but it claims to remove a step in the translation and avoid memory consumption. I could profile the load test to give you numbers on memory because it is not integrated in load tests
Also see my comment https://github.com/apache/beam/pull/11055#discussion_r397246846 regarding materialization.
Anyway there is hotkey configuration in GBKLoadTest. I'll configure it and post the results

Reply:
I just ran this test: `-Xms8g -Xmx8g -Prunner=":runners:spark" -PloadTest.mainClass="org.apache.beam.sdk.loadtests.GroupByKeyLoadTest" -PloadTest.args="--fanout=1 --iterations=1 --streaming=false --runner=SparkStructuredStreamingRunner --sourceOptions={\"numRecords\":200000,\"keySizeBytes\":100000,\"valueSizeBytes\":10000, \"hotKeyFraction\":1.0, \"numHotKeys\":1}"` So one sole key that receives all the values in a 8GO JVM to simulate an OOM.
And I observe a spill to disk and no out of memory as I said:
before run /dev/mapper/ubuntu--vg-root   368G     23G  327G   7% /
after run  /dev/mapper/ubuntu--vg-root   368G     27G  323G   8% /
So +4G on disk in /tmp
runtime_sec                   680.092